lib/Conversion/TorchToMhlo/MatmulOp.cpp

```diff
@@ -0,0 +1,652 @@
+    auto rhsShape = rhsRankTy.getShape();
+    if (lhsRank < rhsRank) {
+      std::vector<int64_t> newShape(rhsShape.begin(),
+                                    rhsShape.begin() + leadingRank);
+      newShape.insert(newShape.end(), lhsShape.begin(), lhsShape.end());
+      auto newDimSizes = *getDimSizesOfTensor(rewriter, op, rhs, leadingDims);
+      auto lhsDimSizes = *getDimSizesOfTensor(rewriter, op, lhs);
+      newDimSizes.insert(newDimSizes.end(), lhsDimSizes.begin(),
+                         lhsDimSizes.end());
+      lhs = getBroadcastTensor(rewriter, op, lhs, newShape, newDimSizes,
+                               broadcastDims);
+    } else {
+      std::vector<int64_t> newShape(lhsShape.begin(),
+                                    lhsShape.begin() + leadingRank);
+      newShape.insert(newShape.end(), rhsShape.begin(), rhsShape.end());
+      auto newDimSizes = *getDimSizesOfTensor(rewriter, op, lhs, leadingDims);
+      auto rhsDimSizes = *getDimSizesOfTensor(rewriter, op, rhs);
+      newDimSizes.insert(newDimSizes.end(), rhsDimSizes.begin(),
+                         rhsDimSizes.end());
+      rhs = getBroadcastTensor(rewriter, op, rhs, newShape, newDimSizes,
+                               broadcastDims);
+    }
+  }
+
+  // [?, ?, m, n] x [?, n, k] ==> batch_matmul([m,n], [n,k])
+  return getDotProduct(rewriter, op, lhs, rhs, /*rank*/ maxRank);
+}
+
+FailureOr<Value> getMmDotProduct(PatternRewriter &rewriter, Operation *op,
+                                 Value inpLhs, Value inpRhs) {
+  auto lhsRankTy = inpLhs.getType().dyn_cast<RankedTensorType>();
+  auto rhsRankTy = inpRhs.getType().dyn_cast<RankedTensorType>();
+
+  auto lhsRank = lhsRankTy.getRank();
+  auto rhsRank = rhsRankTy.getRank();
+  if (lhsRank < 2)
+    return rewriter.notifyMatchFailure(
+        op, "the left hand-side input of matmul must has rank >= 2");
+  if (rhsRank != 2)
+    return rewriter.notifyMatchFailure(
+        op, "the right hand-side input of matmul must has rank == 2");
+
+  Value lhs = inpLhs;
+  Value rhs = inpRhs;
+  // [?, m, n] x [n, k] ==> [?xm, n] x [n, k]
+  std::vector<Value> collapDimSizes;
+  if (lhsRank > 2) {
+    std::vector<int64_t> collapDims;
+    for (int64_t d = 0; d < lhsRank - 1; ++d) {
+      collapDims.push_back(d);
+    }
+    auto collapDimSizesInfo = getCollapsedTensor(rewriter, op, lhs, collapDims);
+    if (failed(collapDimSizesInfo))
+      return rewriter.notifyMatchFailure(
+          op, "failed to construct matrix-matrix multiply");
+    std::tie(lhs, collapDimSizes) = *collapDimSizesInfo;
+  }
+  auto result = getDotProduct(rewriter, op, lhs, rhs, /*rank*/ 2);
+  if (failed(result))
+    return rewriter.notifyMatchFailure(
+        op, "failed to construct matrix-matrix multiply");
+
+  return getExpandedTensor(rewriter, op, *result, collapDimSizes,
+                           /*expandPos*/ 0);
+}
+
+FailureOr<Value> getMvDotProduct(PatternRewriter &rewriter, Operation *op,
```

Comment:
NIT: probably we could use DotGeneralOp directly to avoid unnecessary unsqueeze and reshape operations.

Reply:
Thanks for the advice. I see the implementation of your version in https://github.com/llvm/torch-mlir/pull/1025 is more precise. I had updated the PR to adopt those codes in https://github.com/llvm/torch-mlir/pull/1025.